comet-java-client/src/main/java/ml/comet/experiment/impl/BaseExperiment.java

```diff
@@ -0,0 +1,899 @@
+package ml.comet.experiment.impl;
+
+import io.reactivex.rxjava3.core.Single;
+import io.reactivex.rxjava3.disposables.CompositeDisposable;
+import io.reactivex.rxjava3.functions.Action;
+import io.reactivex.rxjava3.functions.BiFunction;
+import io.reactivex.rxjava3.functions.Function;
+import io.reactivex.rxjava3.schedulers.Schedulers;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.Setter;
+import ml.comet.experiment.Experiment;
+import ml.comet.experiment.exception.CometApiException;
+import ml.comet.experiment.exception.CometGeneralException;
+import ml.comet.experiment.impl.constants.AssetType;
+import ml.comet.experiment.impl.constants.QueryParamName;
+import ml.comet.experiment.impl.http.Connection;
+import ml.comet.experiment.impl.http.ConnectionInitializer;
+import ml.comet.experiment.impl.utils.CometUtils;
+import ml.comet.experiment.model.AddExperimentTagsRest;
+import ml.comet.experiment.model.AddGraphRest;
+import ml.comet.experiment.model.CreateExperimentRequest;
+import ml.comet.experiment.model.CreateExperimentResponse;
+import ml.comet.experiment.model.ExperimentAssetLink;
+import ml.comet.experiment.model.ExperimentMetadataRest;
+import ml.comet.experiment.model.ExperimentStatusResponse;
+import ml.comet.experiment.model.ExperimentTimeRequest;
+import ml.comet.experiment.model.GitMetadata;
+import ml.comet.experiment.model.GitMetadataRest;
+import ml.comet.experiment.model.HtmlRest;
+import ml.comet.experiment.model.LogDataResponse;
+import ml.comet.experiment.model.LogOtherRest;
+import ml.comet.experiment.model.MetricRest;
+import ml.comet.experiment.model.OutputLine;
+import ml.comet.experiment.model.OutputUpdate;
+import ml.comet.experiment.model.ParameterRest;
+import ml.comet.experiment.model.ValueMinMaxDto;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+
+import java.io.File;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static ml.comet.experiment.impl.constants.ApiEndpoints.ADD_ASSET;
+import static ml.comet.experiment.impl.constants.AssetType.ASSET_TYPE_SOURCE_CODE;
+import static ml.comet.experiment.impl.constants.QueryParamName.CONTEXT;
+import static ml.comet.experiment.impl.constants.QueryParamName.EPOCH;
+import static ml.comet.experiment.impl.constants.QueryParamName.EXPERIMENT_KEY;
+import static ml.comet.experiment.impl.constants.QueryParamName.FILE_NAME;
+import static ml.comet.experiment.impl.constants.QueryParamName.OVERWRITE;
+import static ml.comet.experiment.impl.constants.QueryParamName.STEP;
+import static ml.comet.experiment.impl.constants.QueryParamName.TYPE;
+
+/**
+ * The base class for all experiment implementations providing implementation of common routines.
+ */
+public abstract class BaseExperiment implements Experiment {
```

Comment:
How hard would it be to add an equivalent of OfflineExperiment in this architecture?

Reply:
I don't know. Let's create ticket for this and estimate!

Reply:
Ok, not important right now